shared/p2p/service.go

```diff
@@ -78,7 +80,10 @@ type ServerConfig struct {
 // NewServer creates a new p2p server instance.
 func NewServer(cfg *ServerConfig) (*Server, error) {
 	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
I know this a separate issue, but maybe also move this to the point where it is first used, since it would be pointless execution if `buildOptions()` (and other code further down) returns err.

Reply:
Actually now that I read the code a bit more, you're missing a call to `cance()` before you `return nil, err`. Given that you missed that, I'd do a bit of refactoring of the code to not require a call to `cancel()` (or at least add a comment) noting that any early returns added to the codebase need to call `cancel()`. This is makes the code really brittle and prone to easy mistakes like this.

Reply:
good point, will adapt the logic.

Reply:
Please leave this at the top. It is a better pattern to declare the context at the start of the method, especially when referencing tracing spans.

The lost cancel is a memory leak risk and sadly our linter isn't enabled at this time to catch it. There was some problem on mac OS X, I think. So we do need to be mindful of returns where cancel is not called.

Reply:
missed that. Fixing now.